storage/flux/table_test.go

```diff
@@ -1 +1,3549 @@
+	Close           func()
+	influxdb.Reader
+}
+
+func NewStorageReader(tb testing.TB, setupFn SetupFunc) *StorageReader {
+	rootDir, err := ioutil.TempDir("", "storage-flux-test")
+	if err != nil {
+		tb.Fatal(err)
+	}
+
+	var closers []io.Closer
+	close := func() {
+		for _, c := range closers {
+			if err := c.Close(); err != nil {
+				tb.Errorf("close error: %s", err)
+			}
+		}
+		_ = os.RemoveAll(rootDir)
+	}
+
+	// Create a meta client for the storage reader to use.
+	metaConfig := meta.NewConfig()
+	metaConfig.Dir = rootDir
+	metaClient := meta.NewClient(metaConfig)
+	if err := metaClient.Open(); err != nil {
+		close()
+		tb.Fatalf("failed to open meta client: %s", err)
+	}
+	closers = append(closers, metaClient)
+
+	// Create the database.
+	dbName := "db"
+	rpName := "rp"
+
+	// Run the setup function to create the series generator.
+	sg, tr := setupFn(dbName, rpName)
+	rp := &meta.RetentionPolicySpec{
+		Name:               rpName,
+		ShardGroupDuration: 24 * 7 * time.Hour * 52,
+	}
+	if _, err := metaClient.CreateDatabaseWithRetentionPolicy(dbName, rp); err != nil {
+		close()
+		tb.Fatalf("failed to create database: %s", err)
+	}
+
+	// Create the shard group for the data. There should only be one and
+	// it should include the entire time range.
+	sgi, err := metaClient.CreateShardGroup(dbName, rp.Name, tr.Start)
+	if err != nil {
+		close()
+		tb.Fatalf("failed to create shard group: %s", err)
+	} else if sgi.StartTime.After(tr.Start) || sgi.EndTime.Before(tr.End) {
+		close()
+		tb.Fatal("shard data range exceeded the shard group range; please use a range for data that is within the same year")
+	}
+
+	// Create a TSDB for the reader to access.
+	tsdbStore := tsdb.NewStore(rootDir)
+	if err := tsdbStore.Open(); err != nil {
+		close()
+		tb.Fatalf("failed to open TSDB store: %s", err)
+	}
+	closers = append(closers, tsdbStore)
+
+	// Write data into the shard.
+	if err := tsdbStore.CreateShard(dbName, rpName, sgi.Shards[0].ID, true); err != nil {
+		close()
+		tb.Fatalf("failed to create shard: %s", err)
+	}
+	shard := tsdbStore.Shard(sgi.Shards[0].ID)
+	sfile, err := shard.SeriesFile()
+	if err != nil {
+		close()
+		tb.Fatalf("failed to get series file: %s", err)
+	}
+	sfile.DisableCompactions()
+	shardPath := shard.Path()
+	if err := os.MkdirAll(shardPath,0700); err != nil {
+		close()
+		tb.Fatalf("failed to create shard directory: %s", err)
+	}
+
+	if err := writeShard(sfile, sg, sgi.Shards[0].ID, shardPath); err != nil {
+		close()
+		tb.Fatalf("failed to write shard: %s", err)
+	}
+
+	// Run the partition compactor on the series file.
+	for i, p := range sfile.Partitions() {
+		c := tsdb.NewSeriesPartitionCompactor()
+		if err := c.Compact(p); err != nil {
+			close()
+			tb.Fatalf("failed to compact series file %d: %s", i, err)
+		}
+	}
+
+	// Reset the TSDB store.
+	if err := tsdbStore.Close(); err != nil {
+		close()
+		tb.Fatalf("failed to close TSDB store: %s", err)
+	}
+	tsdbStore = tsdb.NewStore(rootDir)
```

Comment:
The previous code hit a race condition when re-openeing the same `tsdbStore`, because a goroutine running in the background tried to read from the "closed" channel while the call to `Open` was re-assigning the channel.